modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/logical/ExposeIndexRule.java

```diff
@@ -71,18 +82,60 @@ public void onMatch(RelOptRuleCall call) {
                 .filter(idx -> filter(igniteTable, idx.indexName(), idx.searchBounds()))
                 .collect(Collectors.toList());
 
+        indexes = applyHints(scan, indexes);
+
         if (indexes.isEmpty()) {
             return;
         }
 
-        Map<RelNode, RelNode> equivMap = new HashMap<>(indexes.size());
+        Map<RelNode, RelNode> equivMap = new HashMap<>(capacity(indexes.size()));
         for (int i = 1; i < indexes.size(); i++) {
             equivMap.put(indexes.get(i), scan);
         }
 
         call.transformTo(indexes.get(0), equivMap);
     }
 
+    /** Filter actual indexes list and prune-table-scan flag if any index is forced to use. */
+    private static List<IgniteLogicalIndexScan> applyHints(TableScan scan, List<IgniteLogicalIndexScan> indexes) {
+        Set<String> tblIdxNames = indexes.stream().map(AbstractIndexScan::indexName).collect(Collectors.toSet());
+        Set<String> idxToSkip = new HashSet<>(capacity(tblIdxNames.size()));
+        Set<String> idxToUse = new HashSet<>(capacity(tblIdxNames.size()));
+
```

Comment:
Is it possible to avoid allocations if there is no hints found?

Reply:
Done